### ISSUE TITLE
Autoencoder w/ Batching: train_test_split_edges error 

### ISSUE THREAD

**Issue by SasCezar** [rusty1s/pytorch_geometric#1817]
## ❓ Questions & Help

I'm trying to implement batching in the autoencoder, but I'm encountering some issues. 
I use MUTAG as dataset:

 ```python
dataset = TUDataset(root='../data/', name='MUTAG',
                    use_node_attr=True, pre_transform=train_test_split_edges)
```

However, when trying to apply the `pre_transform=train_test_split_edges` as you suggest here (Issue #1538), it raises the following error:

```python
  File "/home/cezarsas/anaconda3/envs/GAEClustering/lib/python3.7/site-packages/torch_geometric/datasets/tu_dataset.py", line 144, in process
    self.data, self.slices = self.collate(data_list)
  File "/home/cezarsas/anaconda3/envs/GAEClustering/lib/python3.7/site-packages/torch_geometric/data/in_memory_dataset.py", line 117, in collate
    dim=data.__cat_dim__(key, item))
RuntimeError: Sizes of tensors must match except in dimension 0. Got 17 and 13 in dimension 1 (The offending index is 1)
```

But if I understood correctly, the pre_transofrm function I'm using, should not affect the dimensions of the Data.

**Comment by rusty1s** [comment]
Yeah, the `data.train_neg_adj_mask` cannot be easily batched. This should fix it:
```python
from torch_geometric.datasets import TUDataset
from torch_geometric.utils import train_test_split_edges
import torch_geometric.transforms as T

def custom_transform(data):
    data.train_neg_adj_mask = None
    return data

dataset = TUDataset(
    root='../data/', name='MUTAG', use_node_attr=True,
    pre_transform=T.Compose([train_test_split_edges, custom_transform]))
```

**Comment by SasCezar** [comment]
Thanks, Matthias!

Now I'm trying to understand steps 2 and 3.
When I transform the latent representation _**Z**_ from the encoder into a batch with the following change in the GAE class, it affects the nodes' position; therefore, ` edge_index ` contains outdated information. This then affects the decoding, and more generally the reconstruction loss (as mentioned in suggestion 3).

```python
class GAE(torch.nn.Module):
    ...
        def encode(self, *args, **kwargs):
        r"""Runs the encoder and computes node-wise latent variables."""
        batch = kwargs['batch']
        del kwargs['batch']
        z = self.encoder(*args, **kwargs)
        x, y = to_dense_batch(z, batch=batch) # batch is obtained from the the iteration over the dataloader with data.batch
        return x, y
    ...
```

My question therefore is: is the `to_dense_batch` required?

From the documentation about [Advanced Mini-Batching](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#advanced-mini-batching):
> GNN operators that rely on a message passing scheme do not need to be modified since messages still cannot be exchanged between two nodes that belong to different graphs.
> PyTorch Geometric automatically takes care of batching multiple graphs into a single giant graph with the help of the torch_geometric.data.DataLoader.

I will expect it to be unnecessary to change the GAE, or any other model, since they don't interact with each other. Am I wrong? and if I am, how should the reconstruction process work?

PS: My task is to perform node clustering, and also add a secondary loss on the latent representation.

Thanks for the help.

**Comment by rusty1s** [comment]
The `to_dense_batch` function is need in my opinion to reconstruct the graph since you need a designated batch dimension. This stems from the fact that you do not want reconstruct edges between nodes that are inside different graphs.
However, in case you only train against sparse positive and negative edges, you do not need to convert your latent representation like that. But keep in mind that for negative sampling, you are only interested in negative edges between nodes of the same graph.
